### PR TITLE
This series is a re-rewrite of the image generator code and related manifests

### PR DESCRIPTION
**Pull request by bpradipt**
Closes https://issues.redhat.com/browse/KATA-2633

Current flow :  Create podvm image  -> create mcp, kata rpm etc -> create kata runtimeclass -> create kata-remote runtimeclass
New flow:  create mcp, kata rpm etc -> create kata runtimeclass -> create podvm image  -> create kata-remote runtimeclass

The image generation status events are posted as Kubernetes events for monitoring

How to check for events during image generation when creating kataconfig?
```
oc get events -n openshift-sandboxed-containers-operator --field-selector involvedObject.name=osc-podvm-image-creation
```
How to check for events during image deletion  when deleting kataconfig?
```
oc get events -n openshift-sandboxed-containers-operator --field-selector involvedObject.name=osc-podvm-image-deletion
```

### CODE REVIEW

**Comment by bpradipt** [comment]
Catalog for testing: quay.io/bpradipt/openshift-sandboxed-containers-operator-catalog:v1.5.2

**Comment by bpradipt** [comment]
@vvoronko @snir911

**Comment by bpradipt** [comment]
```
oc get events -n openshift-sandboxed-containers-operator --field-selector involvedObject.name=osc-podvm-image-creation 
LAST SEEN   TYPE     REASON                 OBJECT                         MESSAGE
153m        Normal   PodVMImageJobRunning   job/osc-podvm-image-creation   PodVM image job (osc-podvm-image-creation) is running
155m        Normal   PodVMImageJobRunning   job/osc-podvm-image-creation   PodVM image job (osc-podvm-image-creation) is running
153m        Normal   PodVMImageJobRunning   job/osc-podvm-image-creation   PodVM image job (osc-podvm-image-creation) is running
155m        Normal   PodVMImageJobRunning   job/osc-podvm-image-creation   PodVM image job (osc-podvm-image-creation) is running
155m        Normal   PodVMImageJobRunning   job/osc-podvm-image-creation   PodVM image job (osc-podvm-image-creation) is running
155m        Normal   PodVMImageJobRunning   job/osc-podvm-image-creation   PodVM image job (osc-podvm-image-creation) is running
155m        Normal   PodVMImageJobRunning   job/osc-podvm-image-creation   PodVM image job (osc-podvm-image-creation) is running
155m        Normal   PodVMImageJobRunning   job/osc-podvm-image-creation   PodVM image job (osc-podvm-image-creation) is running
155m        Normal   PodVMImageJobRunning   job/osc-podvm-image-creation   PodVM image job (osc-podvm-image-creation) is running
155m        Normal   PodVMImageJobRunning   job/osc-podvm-image-creation   PodVM image job (osc-podvm-image-creation) is running
148m        Normal   PodVMImageJobRunning   job/osc-podvm-image-creation   PodVM image job (osc-podvm-image-creation) is running
155m        Normal   PodVMImageJobRunning   job/osc-podvm-image-creation   PodVM image job (osc-podvm-image-creation) is running
155m        Normal   PodVMImageJobRunning   job/osc-podvm-image-creation   PodVM image job (osc-podvm-image-creation) is running
155m        Normal   PodVMImageJobRunning   job/osc-podvm-image-creation   PodVM image job (osc-podvm-image-creation) is running
155m        Normal   PodVMImageJobRunning   job/osc-podvm-image-creation   PodVM image job (osc-podvm-image-creation) is running
155m        Normal   PodVMImageJobRunning   job/osc-podvm-image-creation   PodVM image job (osc-podvm-image-creation) is running
155m        Normal   PodVMImageJobRunning   job/osc-podvm-image-creation   PodVM image job (osc-podvm-image-creation) is running
148m        Normal   PodVMImageJobRunning   job/osc-podvm-image-creation   PodVM image job (osc-podvm-image-creation) is running
155m        Normal   PodVMImageJobRunning   job/osc-podvm-image-creation   PodVM image job (osc-podvm-image-creation) is running
155m        Normal   SuccessfulCreate       job/osc-podvm-image-creation   Created pod: osc-podvm-image-creation-q9d76
141m        Normal   Completed              job/osc-podvm-image-creation   Job completed
```

File: controllers/image_generator.go [hunk 1]
```diff
@@ -213,9 +204,19 @@ func (r *ImageGenerator) createJobFromFile(jobFileName string) (*batchv1.Job, er
 		return nil, err
 	}
 
+	// If PODVM_BUILDER_IMAGE environment variable is set, use it
```
Comment (snir911) [inline controllers/image_generator.go:219]: PODVM_BUILDER_IMAGE or RELATED_IMAGE_PODVM_BUILDER?

File: controllers/image_generator.go [hunk 2]
```diff
@@ -225,200 +226,143 @@ func (r *ImageGenerator) createJobFromFile(jobFileName string) (*batchv1.Job, er
...
+	// Set propagation policy to delete all the dependents
+	// This is required to delete the pods created by the job
+	// If the job has already been deleted, return nil
+	// This is to avoid the error "job.batch "jobName" not found"
+	// when the job has already been deleted
 
-	if err := r.client.Delete(context.TODO(), job, &client.DeleteOptions{PropagationPolicy: &deletePolicy}); err != nil {
+	deletePolicy := metav1.DeletePropagationBackground
+	if err := r.client.Delete(context.TODO(), job, &client.DeleteOptions{
+		PropagationPolicy: &deletePolicy,
```
Comment (snir911) [inline controllers/image_generator.go:269]: nit, you can skip setting the deletePolicy variable

File: controllers/image_generator.go [hunk 3]
```diff
@@ -225,200 +226,143 @@ func (r *ImageGenerator) createJobFromFile(jobFileName string) (*batchv1.Job, er
...
-		return false, ctrl.Result{RequeueAfter: 60 * time.Second}
-	} else {
-		igLogger.Info("JobStatus: Unexpected Job Status!!!", "job name", job.Name)
-		if err := r.deleteJobFromFile(filename, true); err != nil {
-			igLogger.Info("error deleting the job definition from yaml file", "err", err, "filename", filename)
-		}
+	err = r.checkJobStatus(job.Name, job.Namespace)
+	if err != nil {
+		igLogger.Info("error checking job status", "err", err)
 		return false, ctrl.Result{Requeue: true}
```
Comment (snir911) [inline controllers/image_generator.go:343]: if job is still running we requeue normally? AFAIR there's some backoff mechanism, is it suitable for a job that we know takes 5-10min ? cc @pmores

Comment (snir911) [reply controllers/image_generator.go:343]: also if error happens, should the job be deleted? IIUC next iteration will fail due to job with the same job name no?

File: controllers/image_generator.go [hunk 4]
```diff
@@ -63,20 +66,33 @@ type ImageGenerator struct {
...
+	igLogger logr.Logger = ctrl.Log.WithName("image-generator")
+)
+
+func InitializeImageGenerator(client client.Client) error {
+	var err error
+	igOnce.Do(func() {
+		ig, err = newImageGenerator(client)
+	})
+	return err
+}
```
Comment (snir911) [inline controllers/image_generator.go:81]: nice:)

File: config/peerpods/podvm/aws-podvm-image-handler.sh [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+    [[ -z "${AWS_SECRET_ACCESS_KEY}" ]] && error_exit "AWS_SECRET_ACCESS_KEY is not set"
+
+    # Packer variables
+    [[ -z "${INSTANCE_TYPE}" ]] && error_exit "INSTANCE_TYPE is not set"
+    [[ -z "${PODVM_DISTRO}" ]] && error_exit "PODVM_DISTRO is not set"
+
+    # if AWS_REGION is empty, try to get it from the metadata service
+    if [[ ! "${AWS_REGION}" ]]; then
+        AWS_REGION=$(curl -m 30 -s --show-error http://169.254.169.254/latest/meta-data/placement/region ||
+            error_exit "Failed to get AWS_REGION from metadata service")
```
Comment (snir911) [inline config/peerpods/podvm/aws-podvm-image-handler.sh:33]: actually if we do this here we can remove the validation of it from the code, however, I'm not sure we should :P

File: controllers/image_generator.go [hunk 6]
```diff
@@ -432,23 +376,279 @@ func (r *ImageGenerator) validatePeerPodsConfigs() error {
...
+			// Event creation is for information purposes, we can just log it
+			err = r.createJobEvent(namespace, jobName, "PodVMImageGenerationNotStarted", fmt.Sprintf("PodVM image generation job %s hasn't started yet", jobName), corev1.EventTypeNormal)
+			if err != nil {
+				igLogger.Info("error creating event for not started Job", "job name", job.Name, "err", err)
+			}
+
+			return fmt.Errorf("job hasn't started yet %s", jobName)
+		}
+	*/
+
```
Comment (snir911) [inline controllers/image_generator.go:605]: not needed?

**Review by snir911 — COMMENTED** [review]
Thanks @bpradipt ! added some comments, do you have a build for this?
If @abhbaner have some cycles for pre-merge test it, it would be perfect (as he found the corner cases in the previous implementation)

File: controllers/image_generator.go (hunk 3 above)
Comment (bpradipt) [reply controllers/image_generator.go]: yes there is an exponential backoff on error. I have updated the code

File: controllers/image_generator.go (hunk 6 above)
Comment (bpradipt) [reply controllers/image_generator.go]: yeah..

**Comment by bpradipt** [comment]
@snir911 catalog for testing - quay.io/bpradipt/openshift-sandboxed-containers-operator-catalog:v1.5.2

File: controllers/image_generator.go (hunk 2 above)
Comment (bpradipt) [reply controllers/image_generator.go:311]: Do we want to keep the pod ?

File: controllers/image_generator.go [hunk 7]
```diff
@@ -105,16 +83,28 @@ func ImageCreate(c client.Client) (bool, ctrl.Result) {
 	}
 
 	if err := ig.validatePeerPodsConfigs(); err != nil {
```
[the remark this answers is not in the document]
Comment (bpradipt) [reply controllers/image_generator.go:129]: It's created

[the remark this answers is not in the document]
Comment (bpradipt) [reply controllers/image_generator.go:129]: createImageConfigMapFromFile

File: controllers/image_generator.go (hunk 1 above)
Comment (bpradipt) [reply controllers/image_generator.go:261]: thanks for the catch

**Review by snir911 — APPROVED** [review]
LGTM overall, assuming pre-merge testing was ok, thanks!

File: config/peerpods/podvm/aws-podvm-image-handler.sh (hunk 5 above)
Comment (bpradipt) [reply config/peerpods/podvm/aws-podvm-image-handler.sh:33]: Yeah, let the validation be there in the code as it will help when we add other image generation methods.

File: config/peerpods/podvm/azure-podvm-image-cm.yaml [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  PODVM_DISTRO: rhel
+
+  # Gallery
+  IMAGE_GALLERY_NAME: "PodVMGallery"
+
+  # Image definition
+  IMAGE_DEFINITION_NAME: "podvm-image"
+  IMAGE_DEFINITION_PUBLISHER: "RedHat"
+  IMAGE_DEFINITION_OFFER: "OSC-PodVM"
+  IMAGE_DEFINITION_SKU: "OSC-PodVM"
```
Comment (snir911) [inline config/peerpods/podvm/azure-podvm-image-cm.yaml:17]: Shouldn't these be set to real defaults?

File: controllers/image_generator.go [hunk 9]
```diff
@@ -73,87 +53,148 @@ const (
 	peerpodsCMAWSImageKey    = "PODVM_AMI_ID"
 	peerpodsCMAzureImageKey  = "AZURE_IMAGE_ID"
 	fipsCMKey                = "BOOT_FIPS"
-	defaultVMType            = "VM"
+	procFIPS                 = "/proc/sys/crypto/fips_enabled"
+	AWSProvider              = "aws"
+	AzureProvider            = "azure"
```
Comment (snir911) [inline controllers/image_generator.go:58]: nit, do you mind adding as part of the series also "peerpodsImageJobsPathLocation" and the other image creation related consts, i think it would make more sense

**Comment by vvoronko** [comment]
new AMI generation flow was pre-merge tested

**Comment by vvoronko** [comment]
I appreciate Pradipta's effort to produce quality code, after few iterations we managed to:
- keep AMI ID at single place
- keep creation and deletion pods (I don't think really, that single pod demands much OCP resources and always could be deleted manually by admin)
- AMI job status provided via kataconfig status is visible in Console
Overall IMHO this improves customer experience

status:
  conditions:
  - lastTransitionTime: "2024-02-22T06:51:26Z"
    message: Creating Pod VM Image
    reason: PodVMImageJobRunning
    status: "True"
    type: InProgress
  kataNodes:
    installed:
    - vvoron415azr22-5hft5-worker-eastus1-8szq5
    nodeCount: 1
    readyNodeCount: 1
  runtimeClasses:
  - kata
  waitingForMcoToStart: false
Screenshot from 2024-02-22 08-53-28

**Comment by bpradipt** [comment]
@gkurz @pmores @littlejawa @jensfr  if you have time, it will be great to have your comments on this PR, especially on the following:

1. New kataconfig status to indicate the pod vm image job creation/deletion
2. I took the liberty to fix some staticcheck warnings and include those as separate commits in the same PR. 


The latest code is pre-merge tested by @vvoronko Thanks a lot for his help. 
Thanks to Snir for initial round of reviews and feedback.

**Comment by vvoronko** [comment]
@jensfr @bpradipt @snir911 Does this PR affect the default AMI image in downstream?

**Comment by bpradipt** [comment]
> @jensfr @bpradipt @snir911 Does this PR affect the default AMI image in downstream?

If you meant whether this PR changes the AMI image that gets created using the existing code, then answer is yes. Since this will use latest pod vm components (like agent, protocol-forwarder etc)
Other than that there are no changes

**Review by littlejawa — APPROVED** [review]
I don't have the environment to test it with, but I trust @vvoronko on that.
LGTM from the code standpoint.

Thanks @bpradipt

**Comment by bpradipt** [comment]
Thanks @littlejawa @snir911 for acks
Thanks @vvoronko for pre-merge testing.
